[executorch][flat_tensor] implement load into and dont hold onto the segment

lucylq opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #8447
* #8437

1. Implement load_into in FlatTensorDataMap
2. Do not persist 'data_ro' in the FlatTensorDataMap. From `get_data`, return the FreeableBuffer given by the data loader.

TODO: add test for load_into.

Differential Revision: [D69148652](https://our.internmc.facebook.com/intern/diff/D69148652/)

lucylq committed:
[executorch][flat_tensor] implement load into and dont hold onto the segment

1. Implement load_into in FlatTensorDataMap
2. Do not persist 'data_ro' in the FlatTensorDataMap. From `get_data`, return the FreeableBuffer given by the data loader.

TODO: add test for load_into.

Differential Revision: [D69148652](https://our.internmc.facebook.com/intern/diff/D69148652/)

[ghstack-poisoned]

lucylq commented:
For: https://github.com/pytorch/executorch/issues/8393

swolchok commented on extension/flat_tensor/flat_tensor_data_map.cpp:
```diff
@@ -97,31 +93,68 @@ ET_NODISCARD Result<FreeableBuffer> FlatTensorDataMap::get_data(
+  // Get metadata to get nbytes.
+  Result<const flat_tensor_flatbuffer::TensorMetadata*> metadata_res =
+      get_flat_tensor_metadata(key, tensor_metadata);
+  if (!metadata_res.ok()) {
+    return metadata_res.error();
+  }
+  const auto metadata = metadata_res.get();
+  ET_CHECK_OR_RETURN_ERROR(
+      metadata->segment_index() >= 0 && metadata->offset() >= 0,
+      InvalidExternalData,
+      "Invalid segment_index %d or offset %lu; malformed PTD file.",
+      metadata->segment_index(),
+      metadata->offset())
+
+  Result<const TensorLayout> tensor_layout = create_tensor_layout(metadata);
+  if (!tensor_layout.ok()) {
+    return tensor_layout.error();
+  }
+  ET_CHECK_OR_RETURN_ERROR(
+      size < tensor_layout.get().nbytes(),
+      InvalidArgument,
+      "Buffer size %zu is smaller than tensor size %zu",
+      size,
+      tensor_layout.get().nbytes())
+
+  const auto* s_data_segment = flat_tensor_->segments();
```
this code is repeated and it's about 20 lines; I would extract a utility function here

lucylq replied:
thanks!

swolchok reviewed (commented):
don't have the context on flat_tensor to accept, here's some minor advice

lucylq committed:
Update on "[executorch][flat_tensor] implement load into and dont hold onto the segment"


1. Implement load_into in FlatTensorDataMap
2. Do not persist 'data_ro' in the FlatTensorDataMap. From `get_data`, return the FreeableBuffer given by the data loader.

TODO: add test for load_into.

Differential Revision: [D69148652](https://our.internmc.facebook.com/intern/diff/D69148652/)

[ghstack-poisoned]

lucylq committed:
Update on "[executorch][flat_tensor] implement load into and dont hold onto the segment"


1. Implement load_into in FlatTensorDataMap
2. Do not persist 'data_ro' in the FlatTensorDataMap. From `get_data`, return the FreeableBuffer given by the data loader.

TODO: add test for load_into.

Differential Revision: [D69148652](https://our.internmc.facebook.com/intern/diff/D69148652/)

[ghstack-poisoned]

lucylq committed:
Update on "[executorch][flat_tensor] implement load into and dont hold onto the segment"


1. Implement load_into in FlatTensorDataMap
2. Do not persist 'data_ro' in the FlatTensorDataMap. From `get_data`, return the FreeableBuffer given by the data loader.

TODO: add test for load_into.

Differential Revision: [D69148652](https://our.internmc.facebook.com/intern/diff/D69148652/)

[ghstack-poisoned]
